libcore/ListModel.vala

```diff
@@ -0,0 +1,323 @@
+    public abstract bool sort_directories_first { get; set; }
+    public abstract ColumnID sort_file_property { get; set; }
+    public abstract bool reversed { get; set; }
+
+    public abstract bool add_file (GOF.File file, GOF.Directory.Async dir);
+    public abstract bool remove_file (GOF.File file, GOF.Directory.Async? dir = null);
+    public abstract bool remove_files (GLib.Sequence<GOF.File> files, GOF.Directory.Async? dir = null);
+    public abstract Gtk.TreeRowReference? find_file_row (GOF.File file, GOF.Directory.Async? dir = null);
+    public abstract GLib.List<Gtk.TreeRowReference> find_file_rows (GLib.Sequence<GOF.File> files, GOF.Directory.Async? dir = null);
+    public abstract GOF.File? file_for_path (Gtk.TreePath path);
+    public abstract GOF.File? file_for_iter (Gtk.TreeIter iter);
+    public abstract bool unload_subdirectory (Gtk.TreeRowReference row_ref);
+
+    /* 'reversed' indicates whether the sort should be the natural order for that property
+     * (defined by the sort function in gof.file) or not */
+    public abstract bool get_order (out FM.ColumnID sort_file_property, out bool reversed);
+    public abstract void set_order (FM.ColumnID sort_file_property, bool? reversed = null);
+}
+
+//    public signal void subdirectory_unloaded (GOF.Directory.Async directory);
+
+public class DirectoryModel : Gtk.TreeStore, DirectoryViewInterface {
+    public GOF.Directory.Async? root_dir { get; set; }
+    public bool has_child { get; set; default = false; }
+    public int icon_size { get; set; default = 32; }
+    public ColumnID sort_file_property { get; set; default = FM.ColumnID.FILENAME;}
+    public bool reversed { get; set; }
+    public bool sort_directories_first { get; set; default = true;}
+    private bool unsorted = false;
+
+    private GLib.HashTable<string, Gtk.TreeRowReference> loaded_subdirectories;
+    construct {
+        loaded_subdirectories = new HashTable<string, Gtk.TreeRowReference> (str_hash, str_equal);
+        set_column_types ({
+              typeof (GOF.File), /* File object */
+            });
+
+        sort_file_property = ColumnID.FILENAME;
+        set_sort_func (ColumnID.FILE_COLUMN, directory_view_sort_func);
+    }
+
+    private int directory_view_sort_func (Gtk.TreeModel model, Gtk.TreeIter iter_a, Gtk.TreeIter iter_b) {
+        if (unsorted) {
+            return 0;
+        }
+
+        GOF.File file_a, file_b;
+        model.@get (iter_a, ColumnID.FILE_COLUMN, out file_a);
+        model.@get (iter_b, ColumnID.FILE_COLUMN, out file_b);
+
+        return file_match_func (file_a, file_b);
+    }
+
+    private int file_match_func (GOF.File? a, GOF.File? b) {
+        if (a == null || b == null) {
+            return 0;
+        }
+
+        return a.compare_for_sort (b, (int)sort_file_property, sort_directories_first, reversed);
+    }
+
+    public new bool get_sort_column_id (out int sort_col, out Gtk.SortType sort_type) {
```

Comment:
I advocate to never use the `new` keyword, either `override` the previous method or use a different name to be sure that the right function will be called.